### ISSUE TITLE
BatchNorm2d doesn't handle inf input in AMP training

### ISSUE THREAD

**Issue by jxtps** [pytorch/pytorch#90342]
### 🐛 Describe the bug

Based on https://discuss.pytorch.org/t/batchnorm2d-outputs-nan-negative-running-var/115373/10

When training a network with AMP, occasional overflow is possible. However, when a BatchNorm2d layer receives `inf` input, it's `state_dict` is corrupted: 
```python
bn = torch.nn.BatchNorm2d(4).to('cuda')
bn.train()
x = torch.ones((1,4, 20, 20)).to('cuda')
x[0, 1, 3, 3] = float('inf')
with torch.autocast(device_type='cuda', dtype=torch.float16, enabled=True):
    y = bn(x)
state_dict = bn.state_dict()
print(state_dict["running_mean"])  # => tensor([0.1000,    inf, 0.1000, 0.1000], device='cuda:0')
print(state_dict["running_var"])  # => tensor([0.9000,    nan, 0.9000, 0.9000], device='cuda:0')
```
The training run is then irretrievably lost since the running stats are used in inference => all `NaN`s, all the time, even though the training loss can recover. 

This type of overflow can happen even for seemingly quite reasonable data & state_dict:
```
Conv2d:
 |   mean   |   min    |   max    |   std    |       shape        || param_name  
 |    0.208 |   -1.560 |    9.680 |    0.997 |    [64, 128, 3, 3] || weight
 |    0.662 |   -2.702 |    8.658 |    1.713 |               [64] || bias
Input: min: 0.0, max: 67.75, mean: 1.1396484375, std: 1.458984375
Output32: min: -2683.9619140625, max: 68526.125, mean: 284.32403564453125, std: 1325.3349609375
Output16: min: -2684.0, max: inf, mean: inf, std: nan
```

Ideally the BatchNorm2d layer would not update its running stats in AMP training with `inf` inputs, so that the scaler can fully roll back the iteration when `inf` is detected. 

### Versions

PyTorch version: 1.12.1+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Home
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.9 (tags/v3.9.9:ccb0e6a, Nov 15 2021, 18:08:50) [MSC v.1929 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19045-SP0
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 516.94
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.23.4
[pip3] torch==1.12.1+cu113
[pip3] torch-tb-profiler==0.4.0
[pip3] torchaudio==0.12.1+cu113
[pip3] torchvision==0.13.1+cu113
[conda] Could not collect

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are @mcarilli @ptrblck @leslie-fang-intel @jgong5

**Comment by jxtps** [comment]
To be clear: this network has historically trained very well under AMP, and this occurred even with gradient clipping.

**Comment by ptrblck** [comment]
> When training a network with AMP, occasional overflow is possible.

Invalid outputs in the forward pass are not expected during mixed-precision training and usually indicate that also your FP32 training would blow up (although it might recover), so I would recommend to try to stabilize your training first.

> so that the scaler can fully roll back the iteration when inf is detected.

The `GradScaler` is responsible for gradient scaling by multiplying the loss with the internal scaling factor. 
Infs in the `.grad` attributes are expected sporadically and the `GradScaler` will then a) skip the `optimizer.step()` call and b) reduce its loss scaling factor. Note that this is different to the overflow during the forward pass, which is not expected as already mentioned. The `GradScaler` thus cannot "rewind" the forward pass.

**Comment by ptrblck** [comment]
> I tried capping the gradients, which didn't really do anything (= it still happened once in a blue moon).

That was a good idea, but I don't think the gradients are at fault based on your description.

> This network has trained well under AMP across many datasets. This is indeed a new-ish dataset, but it's from the same "family" as the others.

That's also interesting to hear, as I would probably start by checking if all input samples are containing valid and "reasonable" values. I've recently debugged a similar issue, where the user eventually realized that the inputs contained some Inf/NaN values and were causing the error.
For your particular use case it would be interesting to see where the invalid values are created in the forward pass, as e.g. also the usage of `torch.log` can easily create them if negative inputs are used. 
A quick way to get some information would be to e.g. use forward hooks and check the activations:
```python
def nan_hook(name):
    def hook(m, input, output):
        if not torch.isfinite(output).all():
            print("Invalid output in {}".format(name))
    return hook

for name, module in model.named_modules():
    module.register_forward_hook(nan_hook(name))
```

**Comment by jxtps** [comment]
The invalid values appear in a Conv2d with seemingly quite reasonable input data & state_dict:
```
Conv2d:
 |   mean   |   min    |   max    |   std    |       shape        || param_name  
 |    0.208 |   -1.560 |    9.680 |    0.997 |    [64, 128, 3, 3] || weight
 |    0.662 |   -2.702 |    8.658 |    1.713 |               [64] || bias
Input: min: 0.0, max: 67.75, mean: 1.1396484375, std: 1.458984375
Output fp32: min: -2683.9619140625, max: 68526.125, mean: 284.32403564453125, std: 1325.3349609375
Output fp16: min: -2684.0, max: inf, mean: inf, std: nan
```
As you can see it's a "minor" overflow - the max output value is 68526.125, which is > 65536.0 so it overflows. 

I've looked at the network's input data (in addition to the input to the specific conv2d cited above) and it looks totally fine, there's nothing that stands out about it. 

On some level it would be fantastic if we could use TF32 (which should really be named fp19, but I digress) across the board. One can dream ;)

**Comment by vadimkantorov** [comment]
Maybe a problem is that batch norm running stats update isn't protected against bad values :( And maybe it also does the update in not high enough accuracy (and overflows?). It's actually a good question: what precision does the update of running stats use?

**Comment by jxtps** [comment]
@ptrblck's reply on https://discuss.pytorch.org/t/batchnorm2d-outputs-nan-negative-running-var/115373/4 asserts: "Batchnorm layers are running in FP32 in an autocast region, so they shouldn’t overflow internally."

**Comment by vadimkantorov** [comment]
I think a part of the problem is that the running_stats update is implicit and currently there's no way to control or cancel it. What may be good is to make this update more explicit, but maybe it's orthogonal to this issue OP. One way hacky way is to clone the running_mean/var into a cache instance variable stored inside the batchnorm module before the training iteration, and then restore them if one wants to cancel the update.

A functional form also allows this somewhat. Although, it may be extended somehow to return the batch mean/var as part of some out-variant or optional variant.

**Comment by jxtps** [comment]
I implemented some BatchNorm-reverting code, but depending on the particulars, a region of the network may just be heading outside of the fp16 bounds. The only way I've found to address this and still use AMP is to disable it for the relevant region, see the last code sample in the docstring for: https://pytorch.org/docs/stable/amp.html#torch.autocast

Unfortunately this leads to a significant slowdown, so I'm not even sure AMP is really worth it in my case.
